### PR TITLE
 enable TCP_NODELAY from config file

### PR DESCRIPTION
**Pull request by wangxufire**

### CODE REVIEW

File: .vscode/c_cpp_properties.json [hunk 1]
```diff
@@ -0,0 +1,92 @@
+{
```
Comment (pexcn) [inline .vscode/c_cpp_properties.json]: .vscode 这堆东西不用 add 进来了吧？

Comment (wangxufire) [reply .vscode/c_cpp_properties.json]: 这些是vscode编辑器的标准配置，可以增加一种开发方式，我觉得挺方便的。

**Comment by greenbambooaaron** [comment]
[*TCP_NODELAY: 2017 Best Practices for TCP Optimization*](
https://www.extrahop.com/company/blog/2016/tcp-nodelay-nagle-quickack-best-practices/)

_Kshama Datar Updated October 12, 2017_

If you want to really understand TCP optimization techniques, how to decide which to use, and how to implement them, you have come to the right place. This post was so long and rich we decided to split it into sections and give it a table of contents. Enjoy!

Table of Contents:
1.TCP background information: Why Nagle's Algorithm and Delayed ACK were implemented and how they interact
Nagle's algorithm and Delayed ACK do not play well together in a TCP/IP network
2. What are TCP_NODELAY and TCP_QUICKACK, and what do they do?
3. Resources on Delayed ACK, Nagle Delays, Tinygrams, Silly Window Syndrome, and other TCP issues
4. Should I enable TCP_NODELAY?
5. How do I figure out if I should enable TCP_NODELAY?
6. How do I know if TCP_NODELAY is helping?
7. How can I resolve the issues caused by Nagle's algorithm and Delayed ACKs?

**Comment by greenbambooaaron** [comment]
Excerpt from the article in the previous post:

"If you are dealing with non-interactive type traffic or bulk transfers such as SOAP, XMLRPC, HTTP/web traffic then enabling TCP_NODELAY to disable Nagle's algorithm is unnecessary.

Some contexts where Nagle's algorithm won't help and TCP_NODELAY should be enabled are:

> Highly interactive applications that communicate with a central server (Citrix, networked video games, etc)
> Telnet-connected devices Applications using chatty protocols (Telnet, SSL)"

**Comment by wangxufire** [comment]
@greenbambooaaron This PR just added a switch in config file, it's implemented by #1685.

**Comment by greenbambooaaron** [comment]
@wangxufire Excellent! Thank you for the extra information. I've been using Shadowsocks for a couple of years now (back in the clowwindy days), but I've only just begun to follow this thread.  Question: If I want to experiment with the nodelay option, how would it appear in the config.json file? "tcp_nodelay"= 1,

And do I have to enable the tcp option on the server too? I'm using Ubu 16.04, so I assume I could do it with sysctl.

Thank you so much for your expertise and great work on this.

**Comment by madeye** [comment]
Please remove all the vscode related changes first.

File: src/tunnel.c [hunk 2]
```diff
@@ -974,6 +974,9 @@ main(int argc, char **argv)
         if (mptcp == 0) {
             mptcp = conf->mptcp;
         }
+        if (no_delay == 0) {
+          no_delay = conf->no_delay;
```
Comment (pexcn) [inline src/tunnel.c]: 这里缩进用 4 个空格

**Comment by madeye** [comment]
Cool!

Merged, thanks!

**Comment by sijidou** [comment]
ss-manager不支持使用--no-delay参数？单独启动倒是没问题。

**Pull request merged**